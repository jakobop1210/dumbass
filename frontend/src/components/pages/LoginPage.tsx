import { FaMusic } from 'react-icons/fa'
import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import InputField from '../atoms/InputField'
import Button from '../atoms/Button'
import { customToast } from '../../lib/utils'
import { LOGIN_USER } from '../../graphql/mutations/userMutations'
import { useMutation } from '@apollo/client'
import { useDispatch } from 'react-redux';
import { setUserLogin, setUserName } from '../../redux/actions/userActions';

/**
 * LoginPage component - Used for user authentication
 *
 * Allows a user to log in by providing a username and a password. After successful login, navigation to the homepage is performed and a successful login message is shown. If the login fails, an error message is displayed.
 *
 * @param {LoginPageProps} props - Properties passed down from parent component. Optionally includes `setUser`.
 */
export default function LoginPage(): JSX.Element {
    const navigate = useNavigate()
    const dispatch = useDispatch()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')
    const [loginUserMutation] = useMutation(LOGIN_USER)

    /**
     * loginUser - Event handler for form submission.
     *
     * Validates provided username and password. If credentials match (username: 'guest', password: 'guest'), navigates to the homepage and optionally invokes the `setUser` function if provided.
     * Shows a toast notification for successful login or failure.
     *
     * @param {React.FormEvent} e - Event object related to the form submission.
     */
    async function loginUser(e: React.FormEvent) {
        e.preventDefault()

        try {
            const result = await loginUserMutation({
                variables: { username, password }
            })
            const { data } = result

            if (data.loginUser) {
                customToast('success', 'Successfully logged in')
                dispatch(setUserLogin()); // Dispatch Redux action to set user as logged in
                dispatch(setUserName(username)); // Dispatch Redux action to set username
               
                // Store username and login status in local storage
                localStorage.setItem('username', username);
                localStorage.setItem('isLoggedIn', 'true');

                navigate('/')
            } else {
                customToast('error', 'An error occurred while logging in')
            }
        } catch (error) {
            console.log(JSON.stringify(error, null, 2))
            customToast('error', 'Wrong password or username')
        }
    }

    return (
        <main className='w-screen flex justify-center'>
            <section className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <header className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <div className='flex items-center justify-center'>
                        <FaMusic size={60} color='#8FC0A9' />
                    </div>
                    <h2 className='mt-10 text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Sign in to your account
                    </h2>
                </header>

                <div className='mt-10 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={loginUser}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <Button title={'Sign in'} type={'submit'} />
                    </form>

                    <footer className='mt-10 text-center text-sm text-blueGray'>
                        Don't have an account?
                        <button
                            onClick={() => navigate('/register')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Register here
                        </button>
                    </footer>
                </div>
            </section>
        </main>
    )
}
