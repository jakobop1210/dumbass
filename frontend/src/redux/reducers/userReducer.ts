import { createReducer } from '@reduxjs/toolkit';
import { setUserLogin, setUserLogout, setUserName } from '../actions/userActions';

interface UserState {
    loggedIn: boolean;
    username: string | null;
}

const initialState: UserState = {
    loggedIn: false,
    username: null,
};

const userReducer = createReducer(initialState, (builder) => {
    builder
        .addCase(setUserLogin, (state) => {
            state.loggedIn = true;
        })
        .addCase(setUserLogout, (state) => {
            state.loggedIn = false;
            state.username = null; // Nullify username
        })
        .addCase(setUserName, (state, action) => {
            state.username = action.payload; // Set username
        });
});

export default userReducer;
