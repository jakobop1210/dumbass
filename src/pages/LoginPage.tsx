import { FaMusic } from 'react-icons/fa'
import { useState } from 'react'
import { useNavigate } from 'react-router-dom'
import toast from 'react-hot-toast'

type LoginPageProps = {
    setUser?: () => void
}

export default function LoginPage({ setUser }: LoginPageProps): JSX.Element {
    const navigate = useNavigate()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')

    function loginUser(e: React.FormEvent) {
        e.preventDefault()

        if (password !== 'guest' || username !== 'guest') {
            toast.error('Wrong password or username', {
                style: {
                    padding: '14px',
                    color: '#696d7d'
                },
                iconTheme: {
                    primary: 'red',
                    secondary: '#FFFAEE'
                }
            })
            return
        }
        if (setUser) setUser()
        navigate('/')
        toast.success('Successfully logged in', {
            style: {
                padding: '14px',
                color: '#696d7d'
            },
            iconTheme: {
                primary: '#8fc0a9',
                secondary: '#FFFAEE'
            }
        })
    }

    return (
        <div className='w-screen flex justify-center'>
            <div className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 mt-10 rounded-lg md:px-20 bg-white'>
                <div className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <div className='flex items-center justify-center'>
                        <FaMusic size={60} color='#8FC0A9' />
                    </div>
                    <h2 className='mt-10 text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Sign in to your account
                    </h2>
                </div>

                <div className='mt-10 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={loginUser}>
                        <div>
                            <label
                                htmlFor='username'
                                className='block text-sm font-medium leading-6 text-blueGray'>
                                Username
                            </label>
                            <div className='mt-2'>
                                <input
                                    id='username'
                                    name='username'
                                    type='username'
                                    autoComplete='username'
                                    placeholder='Username'
                                    value={username}
                                    onChange={(e) =>
                                        setUsername(e.target.value)
                                    }
                                    required
                                    className='block w-full h-12 rounded-md border-0 py-1.5 pl-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red focus:outline-none sm:text-sm sm:leading-6'
                                />
                            </div>
                        </div>

                        <div>
                            <div className='flex items-center justify-between'>
                                <label
                                    htmlFor='password'
                                    className='block text-sm font-medium leading-6 text-blueGray'>
                                    Password
                                </label>
                            </div>
                            <div className='mt-2'>
                                <input
                                    id='password'
                                    name='password'
                                    type='password'
                                    placeholder='Password'
                                    autoComplete='current-password'
                                    value={password}
                                    onChange={(e) =>
                                        setPassword(e.target.value)
                                    }
                                    required
                                    className='block w-full h-12 rounded-md border-0 py-1.5 pl-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:outline-none sm:text-sm sm:leading-6'
                                />
                            </div>
                        </div>

                        <div>
                            <button
                                type='submit'
                                className='flex w-full justify-center rounded-md px-3 py-1.5 text-sm font-semibold leading-8 text-white shadow-[0_4px_9px_-4px_#8FC0A9]'
                                style={{
                                    background:
                                        'linear-gradient(to right, #C8D5B9, #8FC0A9, #68B0AB)'
                                }}>
                                Sign in
                            </button>
                        </div>
                    </form>

                    <p className='mt-10 text-center text-sm text-blueGray'>
                        Don't have an account?
                        <a
                            onClick={() => navigate('/register')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Register here
                        </a>
                    </p>
                </div>
            </div>
        </div>
    )
}
