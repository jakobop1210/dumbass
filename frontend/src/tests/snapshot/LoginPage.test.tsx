import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import { MockedProvider } from '@apollo/client/testing'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import LoginPage from '../../components/pages/LoginPage'
import userReducer from '../../redux/reducers/userReducer'

const mockStore = configureStore({
    reducer: {
        user: userReducer
    },
    preloadedState: {
        user: {
            loggedIn: false,
            username: null
        }
    }
})

test('LoginPage snapshot', () => {
    const { asFragment } = render(
        <Provider store={mockStore}>
            <MockedProvider addTypename={false}>
                <Router>
                    <LoginPage />
                </Router>
            </MockedProvider>
        </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})
