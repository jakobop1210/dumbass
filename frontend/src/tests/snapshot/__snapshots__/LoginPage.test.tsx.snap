// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage snapshot 1`] = `
<DocumentFragment>
  <main
    class="w-screen flex justify-center"
  >
    <section
      class="flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white"
    >
      <header
        class="sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <div
          class="flex items-center justify-center"
        >
          <svg
            color="#8FC0A9"
            fill="currentColor"
            height="60"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(143, 192, 169);"
            viewBox="0 0 512 512"
            width="60"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M470.38 1.51L150.41 96A32 32 0 0 0 128 126.51v261.41A139 139 0 0 0 96 384c-53 0-96 28.66-96 64s43 64 96 64 96-28.66 96-64V214.32l256-75v184.61a138.4 138.4 0 0 0-32-3.93c-53 0-96 28.66-96 64s43 64 96 64 96-28.65 96-64V32a32 32 0 0 0-41.62-30.49z"
            />
          </svg>
        </div>
        <h2
          class="mt-10 text-center text-3xl font-bold leading-9 tracking-tight text-blueGray"
        >
          Sign in to your account
        </h2>
      </header>
      <div
        class="mt-10 sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <form
          class="space-y-6"
        >
          <main>
            <label
              class="block text-sm font-medium leading-6 text-blueGray"
              for="username"
            >
              Username
            </label>
            <input
              class="block w-full h-12 rounded-md border-0 py-1.5 px-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red sm:text-sm sm:leading-6 mt-2"
              id="username"
              name="username"
              placeholder="Username"
              required=""
              type="username"
              value=""
            />
          </main>
          <main>
            <label
              class="block text-sm font-medium leading-6 text-blueGray"
              for="password"
            >
              Password
            </label>
            <input
              class="block w-full h-12 rounded-md border-0 py-1.5 px-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red sm:text-sm sm:leading-6 mt-2"
              id="password"
              name="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
          </main>
          <button
            class="flex w-full justify-center rounded-md px-3 py-1.5 text-sm font-semibold leading-8 text-white shadow-[0_4px_9px_-4px_#8FC0A9] undefined"
            type="submit"
          >
            Sign in
          </button>
        </form>
        <footer
          class="mt-10 text-center text-sm text-blueGray"
        >
          Don't have an account?
          <button
            class="ml-2 font-semibold leading-6 text-green hover:cursor-pointer"
          >
            Register here
          </button>
        </footer>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
